### DISCUSSION TITLE
Support for Typst 0.12.0

### DISCUSSION THREAD

**Discussion by grantmcdermott** [quarto-dev/quarto-cli#11142] [Q&A] [answered]
### Description

Can anyone point me to Quarto's vendoring cycle/policy for Typst?

My question was prompted by the release of Typst 0.12.0 a few days ago, which [includes a bunch of nice features](https://typst.app/blog/2024/typst-0.12/) that I'd like to access via Quarto. In particular, enabling [line numbers](https://typst.app/docs/reference/model/par/#definitions-line).

Thanks!

**Comment by mcanouil** [comment, marked as answer]
It will likely be included in 1.7.

**Reply by grantmcdermott** [reply]
Thanks @mcanouil.

**Reply by cscheid** [reply]
Our general policy with dependencies is:

- ideally, we upgrade at most once every release cycle
- ideally, we upgrade at every release cycle
- ideally, we do not upgrade late in the release cycle

**Reply by danieltomasz** [reply]
I created related   issue https://github.com/quarto-dev/quarto-cli/issues/11294 , but I closed it bc similar issue is  here https://github.com/quarto-dev/quarto-cli/issues/10105, closing mine because it is duplicate,
@grantmcdermott  you might try to use your system typst (for example to have access to bugfix  by setting  shell variable QUARTO_TYPST= to your typst system path)

**Reply by gordonwoodhull** [reply]
FWIW I ran the Quarto tests and have been using Typst 0.12 with Quarto for a week with no issues. 

We need support in Pandoc to support some of its new features, but it seems to be fully compatible. I think we can switch early in 1.7

**Reply by danieltomasz** [reply]
Just a comment about release cycle, after 0.12 typst developers plan to ship new release every 6-8 weeks (0.12 took longer due to architecture changes et cetera) https://forum.typst.app/t/why-is-the-release-cycle-so-slow/570/2 
Quarto has releases around ~ 2 times a year, I  completely understand freezing dependencies in the window before release, given the policies stated above is there  a possible  change  for potentially more dependencies  updates  during the development window if it will be sufficiently early in the development?

**Reply by cscheid** [reply]
Not unless something goes very wrong. We have a lot of dependencies and it's a significant source of toil on our development time.

**Reply by boweihu** [reply]
I am using the newly released Quarto 1.7.2, but the Typst version is still 0.11.0. I'm hoping things get updated and work soon!

**Reply by cderv** [reply]
> I am using the newly released Quarto 1.7.2, but the Typst version is still 0.11.0.

@boweihu yes no update was done yet. You can follow pre-release not to know when this will happen: https://prerelease.quarto.org/docs/download/

**Reply by jeremy886** [reply]
I really want to show emojis in 0.12.0. 🏋️😀

**Comment by grantmcdermott** [comment]
Late follow-up to this, but has anyone successfully managed to include line numbers in the Quarto documents via Typst? I'm struggling, although "native" Typst works as expected.

I've set my `TYPST_QUARTO` shell envar and confirmed with `quarto check` that it's running version 0.12.0.

```sh
quarto check                                                              ─╯
Quarto 1.5.57
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.12.0: OK
...
```

However when I try to render a document with line numbers I encounter an error. Minimal YAML:

```yaml
title: "Important findings"
date: last-modified
format:
  typst:
    include-in-header:
      - text: |
          #set par.line(numbering: "1")
```

Rendering triggers the following error:

```typ
[typst]: Compiling typst-ex.typ to typst-ex.pdf...error: function `par` does not contain field `line`
    ┌─ typst-ex.typ:282:9
    │
282 │ #set par.line(numbering: "1")
```

Grateful for any guidance here...

**Reply by gordonwoodhull** [reply]
Sadly, this is why we can't support nonstandard configurations of Quarto. They may work, but we can't really help you get it working.

In my case I see the expected behavior when I set `QUARTO_TYPST` (please note reversed words in your original message) to the path of the 0.11.1, 0.12, 0.13 Typst binaries.

Your document fails with the error you show when I use Typst 0.11.1, but it succeeds with the later versions.

You can double-check that your Quarto is running the Typst version you expect with

```
quarto typst ---version
```

There shouldn't be any discrepancy with the Typst version reported by `quarto check`, but this is more direct.

**Reply by grantmcdermott** [reply]
> In my case I see the expected behavior when I set QUARTO_TYPST (please note reversed words in your original message) 

Thanks, that was a typo in my message. I used the correct `QUARTO_TYPST` entry in my `.zshrc`. (Sorry for the noise.) For posterity, I ran:

```sh
echo "export QUARTO_TYPST='$(which typst)'" >> ~/.zshrc
```

Where `which typst` gives me the symlink to the 0.12.0 binary (`/usr/local/Cellar/typst/0.12.0/bin/typst`).

> You can double-check that your Quarto is running the Typst version you expect with
> 
> ```
> quarto typst --version
> ```

Still 0.12.0 and having the same rendering problem. I'm stumped, but I'll try on a different OS tomorrow and maybe I'll have more joy. Failing that, looking forward to the 1.7 Quarto release that will bundle Typst v0.12.0 by default ;-)

Appreciate the help.

**Comment by grantmcdermott** [comment]
> Failing that, looking forward to the 1.7 Quarto release that will bundle Typst v0.12.0 by default.

Confirmed that line numbering (and presumably other new features too) work out-of-the-gate with the latest available 1.7.22 pre-release, which [bundles Typst 0.13.0](https://github.com/quarto-dev/quarto-cli/blob/v1.7.22/news/changelog-1.7.md#dependencies). 🎉🎉

Thanks to the @gordonwoodhull and other Quarto-typst devs.

**Reply by roaldarbol** [reply]
Also works for me!

**Comment by roaldarbol** [comment]
As a tangential question, is there a way to specify a path to a different Typst installation? I see there's a `QUARTO_TYPST` global variable. Would it work to set that to point to a different installation? Or would it simply break things?

**Reply by gordonwoodhull** [reply]
Yes, set `QUARTO_TYPST` to the location of a different Typst binary in order to use that instead.

Please note that we can’t support non-standard configurations, so be sure not to have this set when preparing a question or issue.

(But it usually works.)

**Reply by roaldarbol** [reply]
Cool, thanks! Would it only break if I use a template that was designed for an earlier version or is there some stuff happening behind the scenes which can break?

**Reply by gordonwoodhull** [reply]
I can’t answer the first part of your question — that’s up to you to figure out on your own — but no, there is nothing going on behind the scenes. 

We are creating a `.typ` document and then running `typst compile` on it to get a PDF. We might pass a `—font-path` but that’s about it.

**Reply by roaldarbol** [reply]
Coolio, I’ll maybe give it a shot 😅
